eval.c

```diff
@@ -783,31 +811,43 @@ static sexp analyze_var_ref (sexp ctx, sexp x, sexp *varenv) {
   return res;
 }
 
-static sexp analyze_set (sexp ctx, sexp x, int depth) {
+static sexp analyze_set (sexp ctx, sexp x, int depth, int defok) {
   sexp res, varenv;
-  sexp_gc_var2(ref, value);
-  sexp_gc_preserve2(ctx, ref, value);
+  sexp_gc_var4(ref, value, cell, op);
+  sexp_gc_preserve4(ctx, ref, value, cell, op);
   if (! (sexp_pairp(sexp_cdr(x)) && sexp_pairp(sexp_cddr(x))
          && sexp_nullp(sexp_cdddr(x)) && sexp_idp(sexp_cadr(x)))) {
     res = sexp_compile_error(ctx, "bad set! syntax", x);
   } else {
-    ref = analyze_var_ref(ctx, sexp_cadr(x), &varenv);
-    if (sexp_refp(ref) && sexp_lambdap(sexp_ref_loc(ref)))
-      sexp_insert(ctx, sexp_lambda_sv(sexp_ref_loc(ref)), sexp_ref_name(ref));
-    value = analyze(ctx, sexp_caddr(x), depth, 0);
-    if (sexp_exceptionp(ref)) {
-      res = ref;
-    } else if (sexp_exceptionp(value)) {
-      res = value;
-    } else if (sexp_immutablep(sexp_ref_cell(ref))
-               || (varenv && sexp_immutablep(varenv))) {
-      res = sexp_compile_error(ctx, "immutable binding", sexp_cadr(x));
+    cell = sexp_env_cell(ctx, sexp_context_env(ctx), sexp_cadr(x), 0);
+    op = cell ? sexp_cdr(cell) : NULL;
+    if (op && sexp_macrop(op)) {
+      if (!sexp_procedure_variable_transformer_p(sexp_macro_proc(op))) {
+        res = sexp_compile_error(ctx, "can't mutate a syntax keyword", sexp_cadr(x));
+      } else {
+        res = analyze_macro_once(ctx, x, op, depth);
+        if (!sexp_exceptionp(res))
+          res = analyze(ctx, res, depth, defok);
```

Comment:
This means that if a `set!` with a variable-transformer expands into another `set!`, it will add as many stack frames as there are `set!`s, unlike regular macro expansions which can use `goto loop` in the main `analyze` procedure to continue the expansion iteratively. The alternative would be to move at least the checking for the correct form, plus the check for a syntax keyword, into the appropriate clause of `analyze` itself.

Reply:
My two cents: If the macro expansion process so far happens in bounded C stack space, I think it is worth keeping this property.

Reply:
Yes, I agree. Fortunately, a way to do it just occurred to me that neither involves moving half of `analyze_set` into `analyze`, nor using stack space for recursive transformations of variable transformer `set!` operations. I’ll try and implement it tomorrow.

Reply:
Done f63348a